packages/flutter/lib/src/widgets/text.dart

```diff
@@ -713,3 +739,634 @@ class Text extends StatelessWidget {
+    // This loop sends the selection event to one of the following to determine
+    // the direction of the search.
+    //  - currentSelectionEndIndex/currentSelectionStartIndex if the current edge
+    //    is in the current viewport.
+    //  - The opposite edge index if the current edge is not in the current viewport.
+    //  - Index 0 if neither edge is in the current viewport.
+    //
+    // If the result is `SelectionResult.next`, this loop look backward.
+    // Otherwise, it looks forward.
+    //
+    // The terminate condition are:
+    // 1. the selectable returns end, pending, none.
+    // 2. the selectable returns previous when looking forward.
+    // 2. the selectable returns next when looking backward.
+    while (newIndex < selectables.length && newIndex >= 0 && finalResult == null) {
+      currentSelectableResult = dispatchSelectionEventToChild(selectables[newIndex], event);
+      switch (currentSelectableResult) {
+        case SelectionResult.end:
+        case SelectionResult.pending:
+        case SelectionResult.none:
+          finalResult = currentSelectableResult;
+        case SelectionResult.next:
+          if (forward == false) {
+            newIndex += 1;
+            finalResult = SelectionResult.end;
+          } else if (newIndex == selectables.length - 1) {
+            finalResult = currentSelectableResult;
+          } else {
+            forward = true;
+            newIndex += 1;
+          }
+        case SelectionResult.previous:
+          if (forward ?? false) {
+            newIndex -= 1;
+            finalResult = SelectionResult.end;
+          } else if (newIndex == 0) {
+            finalResult = currentSelectableResult;
+          } else {
+            forward = false;
+            newIndex -= 1;
+          }
+      }
+    }
+    if (isEnd) {
+      final bool forwardSelection = currentSelectionEndIndex >= currentSelectionStartIndex;
+      if (forward != null && ((!forwardSelection && forward && newIndex >= currentSelectionStartIndex) || (forwardSelection && !forward && newIndex <= currentSelectionStartIndex))) {
+        currentSelectionStartIndex = currentSelectionEndIndex;
+      }
+      currentSelectionEndIndex = newIndex;
+    } else {
+      final bool forwardSelection = currentSelectionEndIndex >= currentSelectionStartIndex;
+      if (forward != null && ((!forwardSelection && !forward && newIndex <= currentSelectionEndIndex) || (forwardSelection && forward && newIndex >= currentSelectionEndIndex))) {
+        currentSelectionEndIndex = currentSelectionStartIndex;
+      }
+      currentSelectionStartIndex = newIndex;
+    }
+    _flushInactiveSelections();
+    return finalResult!;
+  }
+
+  /// The compare function this delegate used for determining the selection
+  /// order of the selectables.
+  ///
+  /// Defaults to screen order.
+  @override
+  Comparator<Selectable> get compareOrder => _compareScreenOrder;
+
+  int _compareScreenOrder(Selectable a, Selectable b) {
```

Comment:
Could this also be static?

Reply:
Also, how is it that we didn't have logic like this up until this PR? Is there any place to share this logic, especially since it's pretty isolated in static methods?

Anyway, thanks for putting in all of the effort required to get this to work. I think it's not clearly evident that something that sounds simple like adding triple tap support involves bringing all of this massive selection infrastructure to Text as well.

Reply:
It can be static. If you are referring to `_adjustSelection` and `_initSelection`, I am using different implementations compared to the one in the base class https://github.com/flutter/flutter/blob/9d6b4f1c03c507e721860ce638af59505d1e692c/packages/flutter/lib/src/widgets/selectable_region.dart#L1720 . Regarding the `compare*` methods those are also slightly different besides the horizontal and vertical comparison. Also anything below `// From [SelectableRegion].` comes from https://github.com/flutter/flutter/blob/9d6b4f1c03c507e721860ce638af59505d1e692c/packages/flutter/lib/src/widgets/selectable_region.dart#L1560 . I'll evaluate how much of this code we can expose to make the code more shareable in a follow up PR.